subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/SnapshotUtil.java

```diff
@@ -105,4 +105,43 @@ public ReadOnlyFileSystemNode handleUnrelatedToAnyChild() {
         });
         return builder.build();
     }
+
+    private static class SingleChildReadOnlyFileSystemNode implements ReadOnlyFileSystemNode {
+        private final String relativePath;
+        private final FileSystemNode child;
+
+        public SingleChildReadOnlyFileSystemNode(String relativePath, FileSystemNode child) {
+            this.relativePath = relativePath;
+            this.child = child;
+        }
+
+        @Override
+        public Optional<MetadataSnapshot> getSnapshot(VfsRelativePath targetPath, CaseSensitivity caseSensitivity) {
+            return SnapshotUtil.getMetadataFromChildren(getChildren(), targetPath, caseSensitivity, Optional::empty);
+        }
+
+        @Override
+        public boolean hasDescendants() {
+            return child.hasDescendants();
+        }
+
+        @Override
+        public ReadOnlyFileSystemNode getNode(VfsRelativePath relativePath, CaseSensitivity caseSensitivity) {
+            return getChild(getChildren(), relativePath, caseSensitivity);
+        }
+
+        @Override
+        public Optional<MetadataSnapshot> getSnapshot() {
+            return Optional.empty();
+        }
+
+        @Override
+        public Stream<FileSystemLocationSnapshot> rootSnapshots() {
+            return child.rootSnapshots();
+        }
+
+        private SingletonChildMap<FileSystemNode> getChildren() {
+            return new SingletonChildMap<>(relativePath, child);
+        }
```

Comment:
Why create this on-demand instead of storing it as a field?

Reply:
Because we don't need this method right now, since the current code doesn't call a method which needs the children. It probably doesn't matter performance wise whether we store it in a field or use a getter. Do you feel strongly either way?

Reply:
"Because we don't need this method right now" that seems like a code smell. When would we need this method?

Reply:
So we use this method from `DefaultSnapshotHierarchy`. On the returned object, we call either `hasDescendants()` in `hasDescendantsUnder()` or `rootSnapshots()` in `rootSnapshotsUnder()`. That is why we currently don't call any methods on the returned node which need the children.

Though it seems we can get rid of those methods and expose a smaller interface: https://github.com/gradle/gradle/pull/19905